File: 12. Search in BST.py
"""
Task: Search in BST
Time Complexity: O(h)
Space Complexity: O(h) — recursion stack
"""

class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def search_bst(root,key):
    if not root: return False
    if root.data==key: return True
    if key<root.data: return search_bst(root.left,key)
    return search_bst(root.right,key)

if __name__=='__main__':
    root=Node(5); root.left=Node(3); root.right=Node(7)
    print('Search 3:', 'Found' if search_bst(root,3) else 'Not Found')
    print('Search 6:', 'Found' if search_bst(root,6) else 'Not Found')

File: 12. Search in BST.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
bool searchBST(Node* root, int key)
{
    if (!root)
        return false;
    if (root->data == key)
        return true;
    if (key < root->data)
        return searchBST(root->left, key);
    return searchBST(root->right, key);
}
int main()
{
    Node* root = nullptr;
    root = new Node(5);
    root->left = new Node(3);
    root->right = new Node(7);
    cout << "Search 3: " << (searchBST(root, 3) ? "Found" : "Not Found") << endl;
    cout << "Search 6: " << (searchBST(root, 6) ? "Found" : "Not Found") << endl;
    return 0;
}
